RAS-A Exploration Microservice

TSC21 opened the pull request:
Adds the RAS-A Exploration Microservice to the `ras_a` dialect. The Microservice definition will be exposed on the RAS-A IOP.

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <wip/>
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>
+          Sets a specific portal to be an entry portal. Defines also the approach vector for a vehicle to approach and go through the portal.
```
Should be "exit portal"

hamishwillee replied:
Well, the portals should be consistently named. Ingress, Egress or ENTRY/EXIT. I prefer the latter.

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <wip/>
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>
+          Set exploration task world boundaries. When starting a new behavior, either the system has default boundaries or it iss boundless.
```
typo: "iss boundless"

hamishwillee replied:
```suggestion
          Set exploration task world boundaries. When starting a new behavior, either the system has default boundaries or it is boundless.
```

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          Start or continue the exploration task.
+          If starting a new exploration, but requiring to get through a portal first, params 4 or 5 to 7 should be set and different from the specified ignore values.
+          Continuing an exploration can be done without necessarily setting the ingress portal, unless the exploration requires the definition of more than the initial ingress portal.
+          If the passed exploration ID does not exist or is not listed as a valid exploration ID in the vehicle autonomy engine, then this command should be rejected.
```
I'm not sure what this line means. If the passed exploration ID does not exist, doesn't that just mean it's a new exploration?

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -220,5 +382,68 @@
+      <field type="float" name="boundaries_p2_y" units="m" invalid="NaN">Exploration cuboid boundaries point 2 local NED Y coordinate. NaN if unknown.</field>
+      <field type="float" name="boundaries_p3_alt" units="m" invalid="NaN">Exploration cuboid boundaries point 3 Altitude (MSL). NaN if unknown.</field>
+      <field type="float" name="boundaries_p3_z" units="m" invalid="NaN">Exploration cuboid boundaries point 3 local NED Z coordinate. NaN if unknown.</field>
+      <field type="uint32_t" name="ingress_portal_id" invalid="UINT32_MAX">Currently defined ingress portal ID. This portal ID can either be set by the system autonomy engine or by an offboard controller or operator using MAV_CMD_SET_INGRESS_PORTAL. When unknown, not applicable or not deterministic, set to UINT32_MAX.</field>
```
When the DO_EXPLORATION command included a portal, should that populate this as well?

TSC21 replied:
Yes because it is associated to the exploration ID created.

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          cuboid (MAV_FRAME_GLOBAL means global coordinates are being used, while MAV_FRAME_LOCAL_NED means local NED coordinates are being used).
+        </description>
+        <param index="1" label="Exploration task ID" minValue="0" maxValue="255" increment="1">ID of the exploration task to set these boundaries to. Set to UINT8_MAX if not applicable or/and to set these boundaries to the current running or active exploration task.</param>
+        <param index="2" label="Cuboid height" units="m">Exploration 3D space boundaries cuboid height. The Z local coordinate or altitude (MSL) of point 1 and 2 are computed by the sum of this height with the local Z or altitude (MSL) of point 3, i.e Z1 = Z2 = Z3 + cuboid height. NaN if not applicable.</param>
+        <param index="3" label="Approach vector initial point Y or Longitude">Local X (mE4) or Latitude (degE7) (WGS84) of point 1 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="4" label="Approach vector initial point Z or Altitude (MSL)">Local Y (mE4) or Longitude (degE7) (WGS84) of point 1 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="5" label="Approach vector final point X or Latitude">Local X (mE4) or Latitude (degE7) (WGS84) of point 2 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="6" label="Approach vector final point Y or Longitude">Local Y (mE4) or Longitude (degE7) (WGS84) of point 2 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="7" label="Approach vector final point Z or Altitude (MSL)" units="m">Local Z or Altitude (MSL) point 3 of the exploration 3D space boundaries cuboid. This also represents the height of the bottom plane of the cuboid. NaN if not applicable.</param>
```
"Approach vector" doesn't seem applicable, nor do "initial" and "final". Param 4's label and description don't match.

TSC21 replied:
Yes this is a typo. Thanks for checking in

dskyle-shieldai commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        </description>
+        <param index="1" label="Exploration task ID" minValue="0" maxValue="255" increment="1">Sets the ID of the exploration task to start. If the ID already exists, it resumes that (queued/listed) exploration task. Set to UINT8_MAX if one wants to resume the last exploration task.</param>
+        <param index="2" label="Time limit" units="s" minValue="0">Time limit to execute the exploration. Reaching this time triggers the behavior defined in param 4. Set to 0 when there is no time limit.</param>
+        <param index="3" label="Behavior after stopping" minValue="0" maxValue="2" increment="1">The behavior after stopping the exploration task. 0: Hold, 1: Land, 2: Return to position.</param>
```
Should this be "after finishing"? I wouldn't expect this behavior to trigger if the exploration is halted by a mode switch to POSCTL, for example.

dskyle-shieldai commented:
Should we standardize custom modes/sub-modes? Currently, we have ENTER and EXPLORE modes defined, but none of that is in RAS-A.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          COMMAND_INT should be used so to set the MAV_FRAME and consequently, frame and coordinates of the portal position (MAV_FRAME_GLOBAL means global coordinates are being used,
+          while MAV_FRAME_LOCAL_NED means local NED coordinates are being used).
+        </description>
+        <param index="1" label="Portal ID" minValue="0" maxValue="4294967295" increment="1">ID of the portal. If the ID is unknown or not deterministic, set it to UINT32_MAX and use params 5, 6 and 7 to define the local or global position of the portal.</param>
```
What does "not deterministic" mean?

dskyle-shieldai replied:
Also, since this is always a float, UINT32_MAX doesn't make sense. Also, shouldn't the maxValue be the largest integer-safe 32-bit float? Which is 16,777,216 I believe.

hamishwillee replied:
@dskyle-shieldai Good points. If you use the restricted values, can you represent the POIs you might get?

TSC21 replied:
I don't think it matters. For example [here](https://github.com/Dronecode/air-iop-definitions/pull/4/files#diff-48eda1727976f638e3e4752aab8c9ce4818837f66699409fa7ef0f4db4050b79R30-R32), apparently it is possible to concatenate a big-endian 64-bits UID with these two first fields using a 32-bit integer of these fields. I believe it doesn't matter this is defined as a float as long as you cast the fields properly on each side. The only thing it matters is that this is a 32-bit fields. It doesn't necessarily matter the type. Check also [this](https://github.com/mavlink/mavlink/issues/1523).

dskyle-shieldai replied:
What is "casting the fields properly"? Casting an INT32_MAX to a float in the normal way results in a value equivalent to unsigned 0x80000000, which when cast back to an int, is implementation defined, but typically -1.

Are you suggesting that we memcpy integers into the space for the float, then memcpy them back into an int? I'm not sure that's reliable, either: https://stackoverflow.com/questions/57355462/is-it-guaranteed-that-the-copy-of-a-float-variable-will-be-bitwise-equivalent-to

I believe the PR you linked is similarly flawed, and should not take that approach.

hamishwillee replied:
> Are you suggesting that we memcpy integers into the space for the float, then memcpy them back into an int? I'm not sure that's reliable, either: https://stackoverflow.com/questions/57355462/is-it-guaranteed-that-the-copy-of-a-float-variable-will-be-bitwise-equivalent-to

@dskyle-shieldai Yes he is. This approach has been used for 15 years for parameters so perhaps there are cases where it is unreliable in theory, but we "know" it hasn't been a problem in MAVLink.
A description of what happens in this case is in https://mavlink.io/en/services/parameter.html#parameter-encoding

The definition of the [command protocol](https://mavlink.io/en/services/command.html) in mavlink does not address this case - it only states that params 5, 6 may not contain things that can only be sent in floats. 

@TSC21 Obviously you "can" do this. You'd need to document the conversion.
Since you have to split the POI anyway have you considered perhaps it would be easier to split it into smaller chunks that can just be copied into the float directly. You'll lose some spare slots of course, but if you're fairly sure this is stable ...

dskyle-shieldai replied:
Indeed, if that's the intent, it needs to be much more clearly documented. But I don't think it _should_ be the intent. 2^24 portal IDs is plenty for now, and we can contemplate a better solution for integer arguments in general for RAS-A 2.0.

I still believe my suggestion for this spec remains the best available: cap the max at 2^24, which is integer-safe for essentially all implementations of float, and use NaN for the invalid/default.

hamishwillee replied:
I have no strong opinion; either way works, just needs to be unambiguous.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
     <enum name="MAV_CMD">
       <description>Commands to be executed by the MAV. They can be executed on user request, or as part of a mission script. If the action is used in a mission, the parameter mapping to the waypoint/mission message is as follows: Param 1, Param 2, Param 3, Param 4, X: Param 5, Y:Param 6, Z:Param 7. This command list is similar what ARINC 424 is for commercial aircraft: A data format how to interp ...
-      <entry value="199" name="MAV_CMD_SET_POI_FOCUS" hasLocation="false" isDestination="false">
+      <entry value="40500" name="MAV_CMD_GO_THROUGH_PORTAL" hasLocation="true" isDestination="false">
```
What would it do after going through if there is no further instruction? HOld, wait? How far should it go in before it decides to go to the next instruction?

TSC21 replied:
I added a clarification on the description.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>
+          Go through a portal. In an indoor exploration context, a portal represents an structural identifiable entry or pass point to start, stop or continue an exploration.
+          COMMAND_INT should be used so to set the MAV_FRAME and consequently, frame and coordinates of the portal position (MAV_FRAME_GLOBAL means global coordinates are being used,
```
```suggestion
          COMMAND_INT should be used with its MAV_FRAME set appropriately. This will affect the coordinates used for the portal position (MAV_FRAME_GLOBAL means global coordinates are being used,
```

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" reserved="true" default="0"/>
+        <param index="3" reserved="true" default="0"/>
+        <param index="4" reserved="true" default="0"/>
+        <param index="5" label="Position X or Latitude">X or Latitude (WGS84) coordinate of the portal position. NaN or INT32_MAX if unknown (or if using param 1 to set the portal).</param>
```
Maybe
```suggestion
        <param index="5" label="Position X or Latitude">X or Latitude (WGS84) coordinate of the portal position (depending on MAV_FRAME). NaN or INT32_MAX if unknown (or if using param 1 to set the portal).</param>
```

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>
+          Sets a specific portal to be an entry portal. Defines also the approach vector for a vehicle to approach and go through the portal.
+          COMMAND_INT should be used so to set the MAV_FRAME and consequently, frame and coordinates of the approach vector (MAV_FRAME_GLOBAL means global coordinates are being used,
```
This sentence is not well written. See suggestion above https://github.com/Dronecode/air-iop-definitions/pull/2/files#r886226770 - also if you have used the same words elsewhere.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          while MAV_FRAME_LOCAL_NED means local NED coordinates are being used).
+        </description>
+        <param index="1" label="Portal ID" minValue="0" maxValue="4294967295" increment="1">ID of the portal to be set as an ingress point.</param>
+        <param index="2" label="Approach vector initial point X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the initial point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
```
Will this work when only param 5/6 are integers in a COMMAND_INT? This field is a float so
- is using INT32_MAX   safe? - we use NaN normally for floats
- If it is a Lat and you pass into the float for sending, will there be any scaling issues? 

If a problem here, also a problem for param3 and 4.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" label="Approach vector initial point X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the initial point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
+        <param index="3" label="Approach vector initial point Y or Longitude">Y (mE4) or Longitude (degE7) (WGS84) coordinate of the initial point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
+        <param index="4" label="Approach vector initial point Z or Altitude (MSL)" units="m">Z or Altitude (MSL) coordinate of the initial point of the approach vector to the portal. NaN if unknown.</param>
+        <param index="5" label="Approach vector final point X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the final point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
```
You can't set INT32 if this sent in a COMMAND_INT. Ditto for param 6

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" label="Approach vector initial point X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the initial point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
+        <param index="3" label="Approach vector initial point Y or Longitude">Y (mE4) or Longitude (degE7) (WGS84) coordinate of the initial point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
+        <param index="4" label="Approach vector initial point Z or Altitude (MSL)" units="m">Z or Altitude (MSL) coordinate of the initial point of the approach vector to the portal. NaN if unknown.</param>
+        <param index="5" label="Approach vector final point X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the final point of the approach vector to the portal. NaN or INT32_MAX  if unknown.</param>
```
You can't set INT32 if this sent in a COMMAND_INT. Ditto for param 6

TSC21 replied:
Yes I can. Eg. https://mavlink.io/en/messages/common.html#MAV_CMD_DO_ORBIT. I send a NaN for COMMAND_LONG, and INT32_MAX for COMMAND_LONG.

TSC21 replied:
But considering this is a COMMAND_INT (always), what I can't send is a NaN,

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          COMMAND_INT should be used so to set the MAV_FRAME and consequently, frame and coordinates of the approach vector (MAV_FRAME_GLOBAL means global coordinates are being used,
+          while MAV_FRAME_LOCAL_NED means local NED coordinates are being used).
+        </description>
+        <param index="1" label="Portal ID" minValue="0" maxValue="4294967295" increment="1">ID of the portal to be set as an egress point.</param>
```
See all comments for previous message.

TSC21 replied:
Yes I can. Eg. https://mavlink.io/en/messages/common.html#MAV_CMD_DO_ORBIT. I send a NaN for COMMAND_LONG, and INT32_MAX for COMMAND_LONG.

dskyle-shieldai replied:
This is not relevant. The params in question there do change between float and int between COMMAND_LONG and COMMAND_INT. This param does not. It is always a float.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>
+          Start or continue the exploration task.
+          If starting a new exploration, but requiring to get through a portal first, params 4 or 5 to 7 should be set and different from the specified ignore values.
```
What are "specified ignore values"?

hamishwillee replied:
Perhaps you mean
```suggestion
          If starting a new exploration that starts with a portal ingress, then params 4, 5 and 7 must be set with valid portal position information.
```

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <description>
+          Start or continue the exploration task.
+          If starting a new exploration, but requiring to get through a portal first, params 4 or 5 to 7 should be set and different from the specified ignore values.
+          Continuing an exploration can be done without necessarily setting the ingress portal, unless the exploration requires the definition of more than the initial ingress portal.
```
I was about to ask why you need to define the portal position - why not just use the other messages above. It sounds like the case is multiple ingress points that must be navigated before entering a more normal exploration mode?

dskyle-shieldai replied:
It's useful to have an atomic enter-and-explore command, because building entry is a major cause of lost-link. If an entry is commanded separately, first, then the GCS might not be able to command the Explore.

hamishwillee replied:
Assuming you can't know in advance if you'll lose the signal, doesn't that mean you always need to use the enter and explore? - i.e. if you use the enter, and then an expore you're in the same boat. 
So why have the separate enter command?

I'm sure this is bleedingly obvious, but since there are no workflows in the design doc ...

dskyle-shieldai replied:
Just completeness, I think. It's not unreasonable that a future craft might have an "enter portal" feature, but not an "exploration" feature, for example.

hamishwillee replied:
Cool. Mostly I'm just whining because I don't feel I have the full picture, and that makes me uncomfortable.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+          If starting a new exploration, but requiring to get through a portal first, params 4 or 5 to 7 should be set and different from the specified ignore values.
+          Continuing an exploration can be done without necessarily setting the ingress portal, unless the exploration requires the definition of more than the initial ingress portal.
+          If the passed exploration ID does not exist or is not listed as a valid exploration ID in the vehicle autonomy engine, then this command should be rejected.
+          COMMAND_INT should be used so to set the MAV_FRAME and consequently, frame and coordinates of the portal position (MAV_FRAME_GLOBAL means global coordinates are being used,
```
See above. Bad sentence.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" label="Time limit" units="s" minValue="0">Time limit to execute the exploration. Reaching this time triggers the behavior defined in param 4. Set to 0 when there is no time limit.</param>
+        <param index="3" label="Behavior after stopping" minValue="0" maxValue="2" increment="1">The behavior after stopping the exploration task. 0: Hold, 1: Land, 2: Return to position.</param>
+        <param index="4" label="Ingress Portal ID" minValue="0" maxValue="4294967295" increment="1">ID of the ingress portal. If the ID is unknown, not deterministic, or when already in the area to explore after going through a portal, set it to UINT32_MAX and use params 5, 6 and 7 to define the local or global position of the portal. Note that an egress portal is not defined by this command but  ...
+        <param index="5" label="Position X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the portal position. NaN or INT32_MAX if unknown (or if using param 4 to set the portal).</param>
```
As above, how you going to set NaN on an int? Ditto for next param

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" label="Behavior after stopping" minValue="0" maxValue="2" increment="1">The behavior after stopping the exploration task. 0: Hold, 1: Land, 2: Return to position.</param>
+        <param index="3" reserved="true" default="NaN"/>
+        <param index="4" reserved="true" default="NaN"/>
+        <param index="5" reserved="true" default="NaN"/>
+        <param index="6" reserved="true" default="NaN"/>
```
```suggestion
        <param index="5" reserved="true" default="0"/>
        <param index="6" reserved="true" default="0"/>
```
YOu can't default an int to a Nan

TSC21 replied:
If this is sent as a COMMAND_LONG, then I can. But let's make it flexible to be used with both INT and LONG, so let's default to 0.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" reserved="true" default="0"/>
+        <param index="3" reserved="true" default="0"/>
+        <param index="4" label="Yaw or Heading" units="rad">Yaw or heading of the return position. NaN if unknown.</param>
+        <param index="5" label="Position X or Latitude">X (mE4) or Latitude (degE7) (WGS84) coordinate of the return position. NaN or INT32_MAX if unknown.</param>
+        <param index="6" label="Position Y or Longitude">Y (mE4) or Longitude (degE7) (WGS84) coordinate of the return position. NaN or INT32_MAX if unknown.</param>
```
Not Nan for unknown value if this is to be sent in COMMAND_INT

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" reserved="true" default="NaN"/>
+        <param index="3" reserved="true" default="NaN"/>
+        <param index="4" reserved="true" default="NaN"/>
+        <param index="5" reserved="true" default="NaN"/>
+        <param index="6" reserved="true" default="NaN"/>
```
```suggestion
        <param index="5" reserved="true" default="0"/>
        <param index="6" reserved="true" default="0"/>
```

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+      <entry value="40513" name="MAV_CMD_DO_EXPLORATION_RETURN" hasLocation="false" isDestination="false">
+        <wip/>
+        <!-- This command is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>Return to a system-defined position after exploration.</description>
```
Just FMI if sent during when in the middle of the building, is the assumption the vehicle will follow some path or just head straight through the walls, if needed?

dskyle-shieldai replied:
Nova 2 follows a safe path back to the return point. This command is pretty useless otherwise, so perhaps that should be specified.

hamishwillee replied:
I think so. If the safe path back might be controllable in future you might even be specific and/or make this an option. Otherwise you end up having to design it in later, in the say way PX4 support RTL, rally points, mission landing, mission landing reverse (and several more).

Is it a valid assumption that there will always have a safe path back that the vehicle is able to follow? Can you lose positioning data? I'm not saying there isn't  - mostly just interested.

dskyle-shieldai replied:
Of course, you can never assume that anything is safe in all situations. The question is the intent of the command. It should be specified that the intent is that exploration return is safe from indoors, otherwise the command has no purpose.

hamishwillee replied:
Cool. I'm not certain whether we need to say safe path or not - simply because I don't think anyone would assume that the drone would attempt to go through walls. It might be worth saying ("following a system-defined path") - to make it clear that this is not something you can or need to specify.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -10,7 +10,140 @@
+        <param index="2" label="Cuboid height" units="m">Exploration 3D space boundaries cuboid height. The Z local coordinate or altitude (MSL) of point 1 and 2 are computed by the sum of this height with the local Z or altitude (MSL) of point 3, i.e Z1 = Z2 = Z3 + cuboid height. NaN if not applicable.</param>
+        <param index="3" label="Approach vector initial point Y or Longitude">Local X (mE4) or Latitude (degE7) (WGS84) of point 1 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="4" label="Approach vector initial point Z or Altitude (MSL)">Local Y (mE4) or Longitude (degE7) (WGS84) of point 1 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="5" label="Approach vector final point X or Latitude">Local X (mE4) or Latitude (degE7) (WGS84) of point 2 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
+        <param index="6" label="Approach vector final point Y or Longitude">Local Y (mE4) or Longitude (degE7) (WGS84) of point 2 of the exploration 3D space boundaries cuboid. NaN or INT32_MAX if not applicable.</param>
```
As above, the underlying field is an INT.

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -220,5 +382,68 @@
+      <field type="uint8_t" name="exploration_id" invalid="UINT8_MAX">ID of the exploration task. 255 if not applicable or unknown.</field>
+      <field type="uint16_t" name="progress" invalid="UINT16_MAX">Progress measurement of the exploration task. Specific meaning may vary by implementation, but in general, increasing values mean more has been explored. UINT16_MAX when unknown or not applicable.</field>
+      <field type="uint16_t" name="denominator" invalid="0">Measurement of the known size of the exploration task. Specific meaning may vary, but when progress == denominator, this should imply that exploration is complete. This value may increase as more need to explore is discovered, or may be fixed (100 recommended) if the end state is known (e.g., exploration in a known mapped environment). 0 ...
+      <field type="uint8_t" name="flags" enum="MAV_EXPLORATION_STATUS_FLAGS">Bitmap of the exploration task status flags.</field>
```
Are you ever going to need more flags? This field is "full"

hamishwillee commented on message_definitions/v1.0/ras_a.xml:
```diff
@@ -220,5 +382,68 @@
+      <field type="float" name="x" units="m" invalid="NaN">Local X position of this position in the local coordinate NED frame. NaN when unknown.</field>
+      <field type="float" name="y" units="m" invalid="NaN">Local Y position of this position in the local coordinate NED frame. NaN when unknown.</field>
+      <field type="float" name="z" units="m" invalid="NaN">Local Z position of this position in the local coordinate NED frame. NaN when unknown.</field>
+      <field type="float" name="yaw" units="rad" invalid="NaN">World to surface heading transformation of the return-from-exploration position. Used to indicate the heading with respect to the ground. NaN when unknown.</field>
```
NOw you made this just yaw, the description needs to change.

TSC21 replied:
Why? I believe it's the same definition.

hamishwillee replied:
Because when I read "World to surface heading transformation" I don't think "yaw". Who would?

hamishwillee commented:
@TSC21 I've done a first scan based purely on the message definitions. I still don't understand the message flow as per the design doc, so can't tell you if they make sense in context.

hamishwillee commented:
You might consider putting these XML entities **into their own xml** and including it into the ras-a XML.

This isn't a paradigm that is used in MAVLink proper, but I wish it was because it makes it much easier to understand what parts make up a service from the service definition. You can also add and remove them more easily, and see updates.

Just a thought. There are only two issues with this:
- Total nesting can't be more than 5 includes deep.
- It is harder to spot changes in ras-a as they are embedded in sub documents.

TSC21 commented:
@hamishwillee @dskyle-shieldai I believe I have addressed most of your concerns. Details on the implementation will be brought as part of the service definition. The only thing I am unsure is the problem raised with the IDs. I believe that params that the enconding of the params doesn't matter, what matters is that they are 32-bits. Then you just have to address the casting properly on both sides.

TSC21 commented:
Merging this. We can consider the 24-bit IDs for POI's in a further iteration.

@dskyle-shieldai @radarku FYI.
